walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,116 @@
 
 package com.google.sps;
 
+import java.util.List;
+import java.util.Collections;
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.ArrayList;
+import java.util.stream.Collectors;
+
+import com.google.sps.TimeRange;
 
 public final class FindMeetingQuery {
+  private final static int totalMinutes = 1440;
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    //remove events with no requested attendees
+    HashSet<String> requested = new HashSet<>(request.getAttendees());
+    HashSet<Event> eventsSet = new HashSet<>(events);
+
+    ArrayList<TimeRange> mandatoryTimes = eventsSet
+      .stream()
+      .filter(event -> this.checkAttendees(event, request, false) == true)
+      .map(event -> event.getWhen())
+      .collect(Collectors.toCollection(ArrayList::new));
+
+    ArrayList<TimeRange> optionalTimes = eventsSet
+      .stream()
+      .filter(event -> this.checkAttendees(event, request, true) == true)
+      .map(event -> event.getWhen())
+      .collect(Collectors.toCollection(ArrayList::new));
+
+    System.out.println(mandatoryTimes);
+    System.out.println(optionalTimes);
+
+    //sort time ranges by start
+    Collections.sort(mandatoryTimes, TimeRange.ORDER_BY_START);
+    Collections.sort(optionalTimes, TimeRange.ORDER_BY_START);
+
+    ArrayList<TimeRange> mandatoryFreeTimes = findFreeTimes(mandatoryTimes, request);
+    ArrayList<TimeRange> optionalFreeTimes = findFreeTimes(optionalTimes, request);
+
+    if (optionalFreeTimes.isEmpty()) {
+      return mandatoryFreeTimes;
+    }
+    
+    return optionalFreeTimes;
+  }
+
+  private static boolean checkAttendees(Event event, MeetingRequest request, boolean checkOptional) {
+    HashSet<String> attendees = new HashSet<>(event.getAttendees());
+    HashSet<String> requested = new HashSet<>(request.getAttendees());
+    HashSet<String> optional = new HashSet<>(request.getOptionalAttendees());
+
+    for (String person: requested) {
+      if (attendees.contains(person)) {
+        return true;
+      }
+    }
+    
+    if (checkOptional) {
+      for (String person: optional) {
+        if (attendees.contains(person))
+          return true;
+      }
+    }
+
+    return false;
+  }
+
+  private static ArrayList<TimeRange> findFreeTimes(ArrayList<TimeRange> times, MeetingRequest request) {
+    //get free times
+    ArrayList<TimeRange> freeTimes = new ArrayList();
+    
+    if (times.isEmpty() && (request.getDuration() <= totalMinutes)) {
+      freeTimes.add(TimeRange.WHOLE_DAY);
+      return freeTimes;
+    }
+
+    if (times.isEmpty()) {
+      return freeTimes;
+    }
```

Comment:
nitpickiest of nitpicks:

```suggestion
    if (times.isEmpty() && (request.getDuration() <= totalMinutes)) {
      return Array.asList(TimeRange.WHOLE_DAY);
    }

    if (times.isEmpty()) {
      return Collections.emptyList();
    }
    List<TimeRange> freeTimes = new ArrayList();
```

Why like this?

* It is simpler to read and reason (you don't need to remember `freeTimes` being a List.
* It has less possible side effects. If somebody adds a value at the start, we would get weird data.